apiserver/facades/client/charms/services/storage.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2021 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package services
+
+import (
+	"fmt"
+
+	"github.com/juju/charm/v9"
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"github.com/juju/utils/v2"
+
+	charmdownloader "github.com/juju/juju/core/charm/downloader"
+	"github.com/juju/juju/state"
+	stateerrors "github.com/juju/juju/state/errors"
+)
+
+// CharmStorageConfig encapsulates the information required for creating a
+// new CharmStorage instance.
+type CharmStorageConfig struct {
+	// The logger to use.
+	Logger loggo.Logger
+
+	// A factory for accessing model-scoped storage for charm blobs.
+	StorageFactory func(modelUUID string) Storage
+
+	StateBackend StateBackend
+}
+
+// CharmStorage provides an abstraction for storing charm blobs.
+type CharmStorage struct {
+	logger         loggo.Logger
```

Comment:
There's been a pattern emerging to use a Logger interface when mocking the testing, rather than requiring a call to loggo.GetLogger in the tests.

Reply:
I've been following that pattern for the other packages. However, this particular package is meant to handle the plumbing/wiring between the state/logging bits so I opted to use loggo directly here since we also import the state package. 

Due to the `Child` method of the logger, if I were to define an interface here, I would need to define shims at each call site (currently apiserver/charms.go and apiserver/facades/client/charms/client.go but also the new download facade in the future). I don't think there is much benefit from a testing perspective to justify the extra overhead of adding shims for the logger in particular.

Ideally, we should (in a refactoring PR) replace _all_ `loggo.Logger` instances throughout the codebase with a logging interface and add an adapter to the `loggo` package to ensure that the `Child()` call return value is mapped to the logging interface. This way we wouldn't have to use logging shims.

However, if you feel strong about this being an interface let me know and I will push a commit to replace it.